diff --git a/node_modules/lib0/indexeddb.js b/node_modules/lib0/indexeddb.js
index dd4a89a..a9e4888 100644
--- a/node_modules/lib0/indexeddb.js
+++ b/node_modules/lib0/indexeddb.js
@@ -22,7 +22,6 @@ export const rtop = request => promise.create((resolve, reject) => {
   request.onerror = event => reject(new Error(event.target.error))
   /* istanbul ignore next */
   // @ts-ignore
-  request.onblocked = () => location.reload()
   // @ts-ignore
   request.onsuccess = event => resolve(event.target.result)
 })
@@ -45,7 +44,6 @@ export const openDB = (name, initDB) => promise.create((resolve, reject) => {
    */
   request.onerror = event => reject(error.create(event.target.error))
   /* istanbul ignore next */
-  request.onblocked = () => location.reload()
   /**
    * @param {any} event
    */
